File: advent_2015/src/python/advent_2015_day_11.py
import re
input_str = "hxbxwxba"

# part1
def increment(input_sequence: str) -> str:
    if len(input_sequence) == 1:
        i = ord(input_sequence)
        return chr(i+1) if i < 122 else 'a'
    
    next_char = increment(input_sequence[-1])
    if next_char == 'a':
        output_str = increment(input_sequence[:-1]) + 'a'
    else:
        output_str = input_sequence[:-1] + next_char

    return output_str

def valid_pass(input_sequence: str) -> bool:
    consecutive_run = False
    for i in range(1, len(input_sequence) - 1):
        last_char = ord(input_sequence[i-1])
        curr_char = ord(input_sequence[i])
        next_char = ord(input_sequence[i+1])
        if (curr_char == last_char + 1) and (curr_char == next_char - 1):
            consecutive_run = True
    
    forbidden_letters_absent = len(re.findall(r"[iol]", input_sequence)) == 0
    
    double_pattern = len(re.findall(r"(.)\1", input_sequence)) > 1
    
    return consecutive_run and forbidden_letters_absent and double_pattern

def part1():
    output_str = input_str
    while(not valid_pass(output_str)):
        output_str = increment(output_str)
        
    print(output_str)
    
part1()

# part2
part2_input_str = "hxbxxyzz"
def part2():
    output_str = increment(part2_input_str)
    while(not valid_pass(output_str)):
        output_str = increment(output_str)
        
    print(output_str)
    
part2()

File: advent_2015/src/bin/advent_2015_day_11.rs
fn increment(input_sequence: &mut Vec<u8>, index: usize) {
    if index == 0 && input_sequence[index] == 122 {
        return;
    }
    match input_sequence.get(index).unwrap() {
        97..=121 => {
            input_sequence[index] += 1;
        }
        122 => {
            input_sequence[index] = 97;
            increment(input_sequence, index - 1);
        }
        _ => panic!("Invalid input sequence"),
    }
}

fn check_if_valid(input_sequence: &[u8]) -> bool {
    let consecutive_run: bool = input_sequence
        .windows(3)
        .any(|triplet| (triplet[0] + 1 == triplet[1]) && (triplet[1] + 1 == triplet[2]));
    if !consecutive_run {
        return false;
    }

    let forbidden_letters_present: bool = input_sequence
        .iter()
        .any(|&ch| ch == 105 || ch == 111 || ch == 108); // equality with [iol] chars
    if forbidden_letters_present {
        return false;
    }

    let mut double_patterns_count = 0;
    let mut last_double_pattern_idx = 9; // set to be out of array range.
    for i in 1..=7 {
        if (input_sequence[i] == input_sequence[i - 1]) && (i - 1 != last_double_pattern_idx) {
            // second check prevents overlapping patterns like 'aaa'
            double_patterns_count += 1;
            last_double_pattern_idx = i;
        }
    }
    if double_patterns_count < 2 {
        return false;
    }

    true
}

fn part1(input_str: &str) -> String {
    let mut output_str: Vec<u8> = input_str.as_bytes().to_owned();

    while !check_if_valid(&output_str) {
        increment(&mut output_str, 7);
    }

    String::from_utf8(output_str).unwrap()
}

fn part2(input_str: &str) -> String {
    let mut output_str: Vec<u8> = input_str.as_bytes().to_owned();
    increment(&mut output_str, 7);

    while !check_if_valid(&output_str) {
        increment(&mut output_str, 7);
    }

    String::from_utf8(output_str).unwrap()
}

fn main() {
    let input_str: &str = "hxbxwxba";
    let part1_res = part1(input_str);
    let part2_res = part2(&part1_res);

    println!("{} {}", part1_res, part2_res);
}
